### ISSUE TITLE
Rainier include power mode in fan floor determination

### ISSUE THREAD

**Issue by derekhoward55** [ibm-openbmc/dev#1724]
Fan floor based on power mode design needed in both OCC and eBMC.  Eg eBMC code that reads/determines mode put new mode in DBUS, eBMC fan control code read mode and adjust floor accordingly, then mode sent to OCC.  

Whatever other handshaking needed.

**Comment by msbarth** [comment]
Where or who's doing the work to get this power mode on dbus? Also, what is sending the mode to the OCC then?

Seems from a `Rainier Fan Control & Monitoring FF` epic view, only thing needed here is for fan control to get the power mode off dbus and use it appropriately in setting the floor and not the other things listed. Sending the power mode to the OCC sounds like this should be done by the openpower-occ-control app, where it also reads the power mode off dbus and then sends it down to the OCC in its communication path. Then I dont know what code would be reading/determining the power mode and from what, then putting it on dbus....

**Comment by cjcain** [comment]
I am doing the code to put the mode on the dbus and send to the OCC and it will be in in the openpower-occ-control app.
Design doc: https://gerrit.openbmc-project.xyz/c/openbmc/docs/+/38786
Dbus add was merged, but not populated yet: https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-dbus-interfaces/+/42511

**Comment by msbarth** [comment]
@cjcain Awesome, thanks for the info.

**Comment by msbarth** [comment]
@cjcain No, there is no dbus property that represents the fan floor set by fan control. As long as fan control is configured to change the fan floor from a propertiesChanged signal, the new floor will be set soon as fan control gets the signal. If the floor is configured to be changed based on an event timer, then the floor would not be updated until the configured timer expires.

It seems the best solution would be to update the floor based on propertiesChanged signals so that it is updated as soon as fan control gets a signal from a property that affects what the floor should be.

Out of curiosity, what's the reason behind the occ-control app needing to know the floor value? Is this sent down to the OCC for something that it does?

**Comment by cjcain** [comment]
@msbarth occ-control does not need to know the floor value.  It needs to know that the fan code saw the new floor and has at least initiated the floor change.  It is a sequencing issue to maintain thermal safety.

Lets say someone changes from a mode with fan floor of 4000 RPM to a new mode requiring a floor of 8000 RPM.  If OCC made the mode change before the fans were updated, the temperature could rise faster than expected (since the floor was expected to be at 8000).  If the RPM was lower, it is not a big deal because the fan would eventually ramp down.

What would be the longest time it would take from the time dbus is updated to when the fans are updated?  If it could take 5 or more seconds, then we probably need some interface/signal for occ-control to know it is safe to make the mode change.

**Comment by msbarth** [comment]
@cjcain The change in floor (and subsequent increase in fan speeds if the floor is higher than the current target) would happen as soon as fan control got the signal from dbus that the power mode changed. So this would all be dependent on how quickly dbus sends the propertiesChanged signal for the power mode change and fan control is swapped in to process that signal and run the associated actions configured against it (i.e. change in floor). Assuming nothing is bogging down dbus, the change in floor and increase of fan speeds should be within 1sec of the power mode change.

**Comment by derekhoward55** [comment]
Sounds good, lets keep it simple.  If later we see situations where DBus is getting bogged down and the floor isn't getting changed quickly, then we can add some handshaking at that time.  Thanks.

**Comment by msbarth** [comment]
Rainier 4U
- https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-fan-presence/+/50593
- https://gerrit.openbmc-project.xyz/c/openbmc/phosphor-fan-presence/+/50597

Rainier 2U and Everest do not have floor tables utilizing power mode defined by the thermal team yet.

**Comment by derekhoward55** [comment]
No interest from thermal team on needing different values for this.  Closing unless told it's needed.
